mrmustard/physics/representations.py

```diff
@@ -310,3 +308,143 @@ def plot(
+        new._contract_idxs = self._contract_idxs
+        return new
+
+    def __getitem__(self, idx: int | tuple[int, ...]) -> Bargmann:
+        r"""Returns a copy of self with the given indices marked for contraction."""
+        idx = (idx,) if isinstance(idx, int) else idx
+        for i in idx:
+            if i >= self.ansatz.array.dim:
+                raise IndexError(
+                    f"Index {i} out of bounds for ansatz {self.ansatz.__class__.__qualname__} of dimension {self.ansatz.dim}."
+                )
+        new = self.__class__(self.array[idx])
+        new._contract_idxs = idx
+        return new
+
+    def __matmul__(self, other: Fock) -> Fock:
+        r"""Implements the inner product of ansatze across the marked indices.
+
+            Batch:
+            The new Fock will hold the tensor product batch of them.
+            For example:
+                if we have self.array = [batch1,...] and other.array = [batch2,...]
+                the new self.array = [batch1*batch2, ...]
+
+            Order of index:
+            The new Fock's order is arranged as uncontracted elements in self and then other.
+
+        """
+        array1 = self.array
+        array2 = other.array
+        idxs1 = self._contract_idxs
+        idxs2 = other._contract_idxs
+
+        #initial the character list for both objects
+        char1 = np.arange(0,len(array1.shape))
+        char2 = np.arange(len(array1.shape),len(array1.shape)+len(array2.shape))
+
+        #initial the character list for the output object
+        output = [0,len(char1),]
+
+        #update output character list by adding the unused in char1
+        for i,c in enumerate(char1):
+            if i != 0:
+                if not (c in idxs1):
+                    print(c)
+                    output += (c,)
+
+        flag=0
+        for i,c in enumerate(char2):
+            print(i,c)
+            if i in idxs2:
+                #replace the character list in char2 by finding the corresponding in char1
+                char2[i] = char1[idxs1[flag]]
+                flag += 1
+            else:
+                if i!=0:
+                    #update output character list by adding the unused in char2
+                    output += (c,)
+
+        new_array = np.einsum(array1, char1, array2, char2, output)
+        #Rearrange the batch dimension
+        new_shape = [new_array.shape[0] * new_array.shape[1]] + [*new_array.shape[2:]]
+        return self.__class__(array = new_array.reshape(new_shape))
+
+    def trace(self, idxs1: tuple[int, ...], idxs2: tuple[int, ...]) -> Fock:
+        r"""Implements the partial trace over the given index pairs.
+
+        Args:
+            idxs1: indices part1 to trace over
+            idxs2: indices part2 to trace over
+
+        Returns:
+            Fock: the ansatz with the given indices traced over
+        """
+        #TODO: conditions on this idxs1 and idxs2?
+        return np.trace(self.array, axis1=idxs1, axis2=idxs2)
```

Comment:
we can't use numpy, if we do we lose the gradients

Reply:
Needs to fix the trace in the backend to receive the axis as the parameters.